File: webgame/board.js
// Tabuleiro: 1 slot de Monstro + 5 slots de Magia por jogador. Espelha game/board.py.
var TCG = window.TCG || (window.TCG = {});

TCG.N_SLOTS_MAGIA = 5;

TCG.criarLadoTabuleiro = function criarLadoTabuleiro(playerId) {
  return { playerId, monstro: null, magia: new Array(TCG.N_SLOTS_MAGIA).fill(null) };
};

TCG.Board = {
  slotsLivres(lado) {
    const livres = [];
    lado.magia.forEach((v, i) => { if (v === null) livres.push(i); });
    return livres;
  },

  // Próximo slot a ser ocupado quando ninguém pede um índice específico:
  // sempre o mais à DIREITA ainda livre — mesma convenção da mão (ver
  // rectProximoSlotDeMao em ui.js), pra a animação de voo (rectSlotDeMagia)
  // ter garantia de mirar o mesmo lugar onde a carta realmente vai cair.
  proximoSlotLivre(lado) {
    const livres = TCG.Board.slotsLivres(lado);
    return livres.length ? livres[livres.length - 1] : null;
  },

  colocarMonstro(lado, carta) {
    if (lado.monstro !== null) throw new Error("Ja existe um combatente ativo neste lado do tabuleiro.");
    lado.monstro = carta;
  },

  removerMonstro(lado) {
    const carta = lado.monstro;
    lado.monstro = null;
    return carta;
  },

  colocarMagia(lado, carta, slot = null) {
    if (slot === null) {
      slot = TCG.Board.proximoSlotLivre(lado);
      if (slot === null) throw new Error("Nao ha slot de magia livre (limite de 5).");
    } else if (lado.magia[slot] !== null) {
      throw new Error(`Slot de magia ${slot} ja ocupado.`);
    }
    lado.magia[slot] = carta;
    return slot;
  },

  removerMagia(lado, carta) {
    const i = lado.magia.indexOf(carta);
    if (i === -1) throw new Error(`${carta.nome} nao esta em nenhum slot de magia deste lado.`);
    lado.magia[i] = null;
    return i;
  },

  dominioAtivo(lado) {
    return lado.magia.find((c) => c !== null && c.tipo === "Domínio") || null;
  },
};

File: game/board.py
# -*- coding: utf-8 -*-
"""O tabuleiro: por jogador, 1 slot de Monstro (o combatente ativo) e 5 slots
de Magia (Dominio ativo + Maldicoes setadas ocupam slots de magia)."""
from __future__ import annotations

from dataclasses import dataclass, field

N_SLOTS_MAGIA = 5


class SlotOcupadoError(RuntimeError):
    pass


class SlotVazioError(RuntimeError):
    pass


@dataclass
class BoardSide:
    player_id: int
    monstro: int | None = None
    magia: list[int | None] = field(default_factory=lambda: [None] * N_SLOTS_MAGIA)

    def slots_livres(self) -> list[int]:
        return [i for i, v in enumerate(self.magia) if v is None]

    def colocar_monstro(self, card: int) -> None:
        if self.monstro is not None:
            raise SlotOcupadoError("Ja existe um combatente ativo neste lado do tabuleiro.")
        self.monstro = card

    def remover_monstro(self) -> int | None:
        card, self.monstro = self.monstro, None
        return card

    def colocar_magia(self, card: int, slot: int | None = None) -> int:
        if slot is None:
            livres = self.slots_livres()
            if not livres:
                raise SlotOcupadoError("Nao ha slot de magia livre (limite de 5).")
            slot = livres[0]
        elif self.magia[slot] is not None:
            raise SlotOcupadoError(f"Slot de magia {slot} ja ocupado.")
        self.magia[slot] = card
        return slot

    def remover_magia(self, card: int) -> int:
        for i, v in enumerate(self.magia):
            if v == card:
                self.magia[i] = None
                return i
        raise SlotVazioError(f"{card} nao esta em nenhum slot de magia deste lado.")

    def dominio_ativo(self, world, dominio_cards: set[int]) -> int | None:
        """Ajuda a achar qual carta nos slots de magia e o Domínio ativo
        (so pode haver 1 por vez, GAME_DESIGN.md — Fase Principal)."""
        for v in self.magia:
            if v is not None and v in dominio_cards:
                return v
        return None


@dataclass
class Board:
    lados: dict[int, BoardSide] = field(default_factory=dict)

    def lado(self, player_id: int) -> BoardSide:
        if player_id not in self.lados:
            self.lados[player_id] = BoardSide(player_id=player_id)
        return self.lados[player_id]
